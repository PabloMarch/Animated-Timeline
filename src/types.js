// @flow

// Animation engine instance
export type AnimationEngine = Object

// Animation attributes
export type attributes = Object

// Controls for time-based execution (start, stop, reverse, reset, restart)
export type ctrl = () => void

// Time-based execution methods
export type controller = {
  start: ctrl,
  stop: ctrl,
  restart: ctrl,
  reverse: ctrl,
  reset: ctrl,
}

// Props dispatched to lifecycle hooks
export type lifecycleHookProps = {
  completed: boolean,
  progress: number,
  duration: number,
  remaining: number,
  reversed: boolean,
  currentTime: number,
  began: boolean,
  paused: boolean,
  controller: controller,
}

// Lifecycle hook signature
export type fn = (props: lifecycleHookProps) => void

// Animation lifecycle hooks
export type lifecycle = {
  onStart: fn,
  tick: fn,
  onUpdate: fn,
  onComplete: fn,
}
