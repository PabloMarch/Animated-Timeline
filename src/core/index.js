import { bezier } from './bezier'
import { easings } from './easing'
import {
  animated,
  getValue,
  createTimeline,
  random
} from './engine'

export {
  bezier,
  easings,
  animated,
  getValue,
  createTimeline,
  random
}