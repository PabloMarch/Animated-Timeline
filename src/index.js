import Engine from 'engine-fork'

import { createCurve } from './utils/createCurve'
import { hx } from './utils/nameToHex'
import { Keyframes } from './utils/keyframes'
import { start, times, startAfter, startBefore } from './utils/properties'
import { Playback } from './components/Playback'
import { Timeline } from './components/Timeline'

export const Animated = {
  createCurve,
  hx,
  start,
  times,
  startAfter,
  startBefore,
  Playback,
  Timeline,
  Keyframes,
  velocity: Engine.speed,
  suspend: Engine.remove
}
