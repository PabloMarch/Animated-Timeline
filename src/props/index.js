// @flow

import PropTypes from "prop-types";

import { noop } from "../utils/noop";

import type { seekCtrl } from "../types";

export const defaultCommonProps = () => ({
  // controls for time-based execution
  start: false,
  stop: false,
  reverse: false,
  reset: false,
  restart: false,

  seek: (ctrl: seekCtrl) => ctrl.default(1),

  // Animation lifecycle
  lifecycle: {
    onUpdate: noop,
    onStart: noop,
    onComplete: noop,
    callFrame: noop
  }
});

export const defaultPropTypes = () => ({
  start: PropTypes.bool,
  stop: PropTypes.bool,
  reverse: PropTypes.bool,
  reset: PropTypes.bool,
  restart: PropTypes.bool,

  seek: PropTypes.func,

  lifecycle: PropTypes.shape({
    onUpdate: PropTypes.func,
    onStart: PropTypes.func,
    onComplete: PropTypes.func,
    callFrame: PropTypes.func
  })
});
