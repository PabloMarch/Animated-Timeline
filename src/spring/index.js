const rebound = require('rebound')

function createSpring(friction, tension) {
  const springSystem = new rebound.SpringSystem()
  let spring = springSystem.createSpring(friction, tension)
  let id = null

  spring.init = (prevState, el, prop, options) => {
    // el.style.transform = prop + '(' + prevState + ')'

    spring.addListener({
      onSpringUpdate: spr => {
        var val = spr.getCurrentValue()
        // val = rebound.MathUtil.mapValueInRange(val, from[0], from[1], to[0], to[1]);
        if (prop === 'rotate') val = String(val).concat('deg')

        id = window.requestAnimationFrame(
          () => (el.style.transform = prop + '(' + val + ')')
        )
      },
    })
  }

  spring.stop = () => id && window.cancelAnimationFrame(id)

  spring.setValue = spring.setEndValue

  return spring
}

class App extends Component {
  state = {
    value: 1,
  }

  componentDidMount() {
    spring.init(this.state.value, '#one', 'scale', {
      mapValues: {
        from: [],
        to: []
      },
      interpolateColors: {
        value: 1,
        startColor: 'red',
        endColor: 'blue',
        from: [0, 200],
        rgb: false,
      }
    })
  }

  render() {
    return (
      <div
        id="one"
        onMouseUp={() => spring.setValue(4)}
        onMouseDown={() => spring.setValue(3)}
      >
        Hey, Nitin!
      </div>
    )
  }
}

// Spring system for the animatione engine

const spring = createSpring(friction, tension)

const mapValues = {
  from: [0, 1],
  to: [0, 0.5]
}

const interpolateColors = {
  value: 1,
  startColor: 'red',
  endColor: 'blue',
  from: [0, 200],
  rgb: false,
}

componentDidMount() {
  spring.init(
    this.state.value, // Initial value for the animation property (scale(1))
    element, // refs or id or classname
    properties, // 'scale'
    { mapValues, interpolateColors }
  )
}

handleUp() {
  spring.setValue(5)
}

handleDown() {
  spring.setValue(3)
}

<div id="hello-world" onMouseUp={handleUp} onMouseDown={handleDown}>
  Hello World
</div>
