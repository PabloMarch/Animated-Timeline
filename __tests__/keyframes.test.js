import { Keyframes } from '../src'

describe('Keyframes', () => {
  it('Should return array of frames', () => {
    const { frames } = new Keyframes()
      .add({
        value: 500,
        duration: 3000,
        elasticity: 900,
      })
      .add({ value: 1000, duration: 1000 })
      .add({
        value: 500,
        duration: 3000,
        elasticity: 900,
      })
      .add({ value: 1000, duration: 1000 })

    expect(Array.isArray(frames)).toBe(true)
    expect(frames).toMatchSnapshot()
  })
})
