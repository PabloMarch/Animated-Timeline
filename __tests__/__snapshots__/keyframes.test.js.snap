// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyframes Should return array of frames 1`] = `
Array [
  Object {
    "duration": 3000,
    "elasticity": 900,
    "value": 500,
  },
  Object {
    "duration": 1000,
    "value": 1000,
  },
  Object {
    "duration": 3000,
    "elasticity": 900,
    "value": 500,
  },
  Object {
    "duration": 1000,
    "value": 1000,
  },
]
`;
